File: BOJ/3980.cpp
#include <stdio.h>
#include <algorithm>
#include <cstring>
using namespace std;
int t,ans;
int board[13][13];
bool visited[13];
void go(int index, int cnt,int hab){
	if (index > 12) return;
	if (cnt == 11) {
		ans = max(ans, hab);
		return;
	}
	for (int i = 1; i <= 11; i++){
		if (board[index][i] == 0) continue;
		if (!visited[i]){
			visited[i] = true;
			hab += board[index][i];
			go(index + 1, cnt + 1, hab);
			visited[i] = false;
			hab -= board[index][i];
		}
	}
}
int main() {
	scanf("%d", &t);
	while (t--) {
		ans = 0;
		memset(visited, false, sizeof(visited));
		for (int i = 1; i <= 11; i++) {
			for (int j = 1; j <= 11; j++) {
				scanf(" %d", &board[i][j]);
			}
		}
		go(1,0,0);
		printf("%d\n", ans);
	}
	return 0;
}

File: BOJ(Java 공부용)/3980.java
import java.io.*;
import java.util.StringTokenizer;
import static java.lang.Integer.max;
public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int[] gox = {0, 1, -1, 0, 1, 1, -1, -1};
    static int[] goy = {1, 0, 0, -1, 1, -1, 1, -1};
    static int ans = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            int[][] board = new int[11][11];
            boolean[] visited = new boolean[11];
            for (int i = 0; i < 11; i++) {
                StringTokenizer st = new StringTokenizer(br.readLine());
                for (int j = 0; j < 11; j++) {
                    board[i][j] = Integer.parseInt(st.nextToken());
                }
            }
            ans = 0;
            go(0, 0, 11, 0, board, visited);
            bw.write(ans + "\n");
        }
        bw.flush();
        bw.close();
    }

    private static void go(int index, int cnt, int depth, int sum, int[][] board, boolean[] visited) {
        if (cnt == depth) {
            ans = max(ans, sum);
            return;
        }
        for (int i = 0; i < 11; i++) {
            if (board[index][i] == 0) continue;
            if (!visited[i]) {
                visited[i] = true;
                go(index + 1, cnt + 1, depth, sum + board[index][i], board, visited);
                visited[i] = false;
            }
        }
    }
}